components/secret-mgt/org.wso2.carbon.identity.secret.mgt.core/src/main/java/org/wso2/carbon/identity/secret/mgt/core/IdentityProviderSecretsProcessorImpl.java

```diff
@@ -0,0 +1,133 @@
+import org.wso2.carbon.identity.secret.mgt.core.model.Secret;
+import org.wso2.carbon.identity.secret.mgt.core.model.SecretType;
+import static org.wso2.carbon.identity.secret.mgt.core.constant.SecretConstants.IDN_SECRET_TYPE_IDP_SECRETS;
+
+/**
+ * Identity provider secrets processor service implementation.
+ */
+public class IdentityProviderSecretsProcessorImpl implements IdentityProviderSecretsProcessor {
+
+    private final SecretManager secretManager;
+    private final SecretResolveManager secretResolveManager;
+
+    public IdentityProviderSecretsProcessorImpl() {
+
+        this.secretManager = new SecretManagerImpl();
+        this.secretResolveManager = new SecretResolveManagerImpl();
+    }
+
+    @Override
+    public IdentityProvider addOrUpdateIdPSecrets(IdentityProvider identityProvider) throws SecretManagementException {
+
+        for (FederatedAuthenticatorConfig fedAuthConfig : identityProvider.getFederatedAuthenticatorConfigs()) {
+            for (Property prop : fedAuthConfig.getProperties()) {
+                if (prop.isConfidential()) {
+                    String secretName = buildSecretName(identityProvider.getId(), fedAuthConfig.getName(), prop.getName());
+                    if (secretManager.isSecretExist(IDN_SECRET_TYPE_IDP_SECRETS, secretName)) {
+                        // Update existing secret property.
+                        updateExistingSecretProperty(secretName, prop);
+                        prop.setValue(buildSecretReference(secretName));
+                    } else {
+                        // Add secret to the DB.
+                        if (!StringUtils.isEmpty(prop.getValue())) {
+                            addNewIdpSecretProperty(secretName, prop);
+                            prop.setValue(buildSecretReference(secretName));
+                        }
+                    }
+                }
+            }
+        }
+
+        return identityProvider;
+    }
+
+    @Override
+    public IdentityProvider getIdPSecrets(IdentityProvider identityProvider)
+            throws SecretManagementException {
+
+        for (FederatedAuthenticatorConfig fedAuthConfig : identityProvider.getFederatedAuthenticatorConfigs()) {
+            for (Property prop : fedAuthConfig.getProperties()) {
+                if (prop.isConfidential()) {
+                    String secretName = buildSecretName(identityProvider.getId(), fedAuthConfig.getName(), prop.getName());
+                    if (secretManager.isSecretExist(IDN_SECRET_TYPE_IDP_SECRETS, secretName)) {
+                        ResolvedSecret resolvedSecret =
+                                secretResolveManager.getResolvedSecret(IDN_SECRET_TYPE_IDP_SECRETS, secretName);
+                        // Replace secret reference with decrypted original secret.
+                        prop.setValue(resolvedSecret.getResolvedSecretValue());
+                    }
+                }
+            }
+        }
+
+        return identityProvider;
```

Comment:
We need to return a new object. changing original object will cause problems if someone saves the original object back to the DB as the original object now has the unencrypted values.

Reply:
Resolved with [4cca0ec](https://github.com/wso2/carbon-identity-framework/pull/4430/commits/4cca0ec26dfa11c945d145f1a0ba1ad3d2cabd20)